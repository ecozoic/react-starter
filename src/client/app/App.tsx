import * as React from 'react';
import { History } from 'history';
import * as PropTypes from 'prop-types';
import { ConnectedRouter } from 'connected-react-router';

import routes from './routes';

interface AppProps {
  history: History;
}

/**
 * Main application component.
 */
class App extends React.Component<AppProps, undefined> {
  static propTypes = {
    history: PropTypes.object.isRequired,
  };

  render() {
    const { history } = this.props;

    return (
      <ConnectedRouter history={history}>
        {routes}
      </ConnectedRouter>
    );
  }
}

export default App;
