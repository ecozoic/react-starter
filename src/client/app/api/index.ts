import { Observable } from 'rxjs/Observable';

const { fetch } = window;
const API_URL = '/api/todos';

export const getTodos: () => Observable<string[]> = () => {
  return Observable.fromPromise(
    fetch(API_URL).then((response) => {
      if (!response.ok) {
        throw Error(`${response.status}: ${response.statusText}`);
      }

      return response.json();
    }),
  );
};
