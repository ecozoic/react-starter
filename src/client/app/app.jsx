/* @flow */
import React, { Component, PropTypes } from 'react';
import { ConnectedRouter } from 'connected-react-router/immutable';

import { routes } from './routes';

/**
 * Main application component.
 * @extends React.Component
 */
export class App extends Component {
  static propTypes = {
    history: PropTypes.object.isRequired
  }

  /**
   * Render method.
   */
  render() {
    const { history } = this.props;

    return (
      <ConnectedRouter history={history}>
        { routes }
      </ConnectedRouter>
    );
  }
}
