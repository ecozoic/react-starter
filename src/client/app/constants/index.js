/* @flow */
export const ADD_TODO = 'ADD_TODO';
export const TOGGLE_TODO = 'TOGGLE_TODO';

export const FETCH_TODOS = 'FETCH_TODOS';
export const FETCH_TODOS_PENDING = 'FETCH_TODOS_PENDING';
export const FETCH_TODOS_FULFILLED = 'FETCH_TODOS_FULFILLED';
export const FETCH_TODOS_REJECTED = 'FETCH_TODOS_REJECTED';
