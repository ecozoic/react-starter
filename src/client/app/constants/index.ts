export const ADD_TODO = 'ADD_TODO';
export const TOGGLE_TODO = 'TOGGLE_TODO';

export const FETCH_TODOS = 'FETCH_TODOS';
export const FETCH_TODOS_PENDING = `${FETCH_TODOS}_PENDING`;
export const FETCH_TODOS_FULFILLED = `${FETCH_TODOS}_FULFILLED`;
export const FETCH_TODOS_REJECTED = `${FETCH_TODOS}_REJECTED`;
