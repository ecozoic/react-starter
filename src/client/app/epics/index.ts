import { combineEpics } from 'redux-observable';

import { fetchTodosEpic } from './todos';

const rootEpic = combineEpics(
  fetchTodosEpic,
);

export default rootEpic;
