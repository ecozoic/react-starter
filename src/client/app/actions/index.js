/* @flow */
import * as t from '../constants';

let nextTodoId = 0;

export const addTodo = (text) => {
  return {
    type: t.ADD_TODO,
    id: nextTodoId++,
    text
  };
};

export const toggleTodo = (id) => {
  return {
    type: t.TOGGLE_TODO,
    id
  };
};
