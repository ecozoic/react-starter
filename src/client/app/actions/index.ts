import { ADD_TODO, TOGGLE_TODO, FETCH_TODOS } from '../constants';

const { fetch } = window;

let nextTodoId = 0;

export const addTodo = (text: string) => ({
  type: ADD_TODO,
  payload: {
    // tslint:disable-next-line:no-increment-decrement
    id: nextTodoId++,
    text,
  },
});

export const toggleTodo = (id: number) => ({
  type: TOGGLE_TODO,
  payload: {
    id,
  },
});

export const fetchTodos = () => (dispatch: Function) => dispatch({
  type: FETCH_TODOS,
  payload: fetch('/api/todos')
    .then((response) => response.json())
    .then((todos: string[]) => {
      todos.forEach((todo) => dispatch(addTodo(todo)));
    }),
});
