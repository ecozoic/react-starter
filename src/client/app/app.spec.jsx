/* @flow */
import React from 'react';
import { shallow } from 'enzyme';

import { App } from './app';

test('it works', () => {
  const component = shallow(
    <App />
  );

  expect(2).toEqual(2);
});
