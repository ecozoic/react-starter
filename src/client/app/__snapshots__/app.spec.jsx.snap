exports[`App renders without crashing 1`] = `
<Connect(ConnectedRouter)
  history={Object {}}>
  <div>
    <Switch>
      <Route
        component={[Function]}
        exact={true}
        path="/" />
      <Route
        component={[Function]} />
    </Switch>
  </div>
</Connect(ConnectedRouter)>
`;
